### DISCUSSION TITLE
Inferring a literal string type

### DISCUSSION THREAD

**Discussion by decorator-factory** [microsoft/pyright#3589] [Ideas]
Right now this code causes an error in Pyright, at the last line:
```py
from typing import TypeVar, Generic
from typing_extensions import LiteralString

L = TypeVar("L", bound=LiteralString)

class Foo(Generic[L]):
    def __init__(self, value: L) -> None:
        self.value = value

foo = Foo("hmm")
```
```
Argument of type "Literal['hmm']" cannot be assigned to parameter "value" of type "L@Foo" in function "__init__"
  Type "str" cannot be assigned to type "LiteralString"
    "str" is incompatible with "LiteralString"
```
I think it would be useful if `L` would be solved as `Literal["hmm"]` here.

My use case is: I'm making a parser combinator library. It would be useful if something like this was possible:
```py
# const: L -> Parser[L]

user_type = const("admin") | const("user") | const("bot")
# user_type: Parser[Literal["admin", "user", "bot"]]
```

Another use case: pathlib-style HTTP routes with parameters that remember which parameters are valid.
```py
get_image_path = Route() / "users" / Var("user_id") / "images" / Var("image_id")
# image_path: Route[Literal["user_id", "image_id"]]

@framework.get(get_image_path)
def get_image(req: Request[Literal["user_id", "image_id"]]) -> Response:
    req.param("user_id")  # ok
    req.param("userId")  # error
    ...
```

I got the idea from this being possible in TypeScript: [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAShDOBXANsAPAFQHxQLwCgoioAfKAbwHsBrALimACdEIAaKRhYe+JgSwB2Ac3YA3AIbIW9DAF9CxMlTpQAZpPhsoERo0qN64gSHn5QkKAAVxjTY0w5cUABTxKiRgGMIPfsICUeDhwSKgO+BGelAK8UMh8wDqSeFBoAMraAB6JAgAm8FC8jIJCWM4SyPRp-vTWtjrpjjhuHt5BCkQtXtC4vVAVHVAA-BQ09Ews7Jy89BUAdPDx3uWSc8gQwsAAFv5iktL9kvLEUPTK9OrImuw6egYUEJmQnom5s5LsQpTche7dsqYojFgFB9Ck0ABBdgAITKYHEtRsdkhWHYYAARoj6vZYTUrEiGhDSFBYe0Tl02rgcORBicgbF4fVckSnPDXH9vP5acQ+KoXIzNMy5jQuScxcROMAPAIoAKIMzuUR6SC5bloSkMezWhBReKoLz+QS1cLqLq9RKIFLGDLVdDFRxLdLRipLtdtLp9PQANqqiFzW76NFG6H+j2MAC6x2IpnwyqgiDsGHAPVBjGc+mc8USjEkzgAROJcgBbQR53ZxBJJZD5hM6MvlrNV-Po771-AAenbUC2lFEOlBfcYUAABrXGEnIMP8EA)

**Comment by erictraut** [comment]
I agree this is a bug. It will be fixed in the next release.

The constraint solver includes some heuristics to determine whether it should retain literal values when it's inferring type variable values from arguments. The heuristic didn't take into account `LiteralString`. Now it does. :)

**Reply by decorator-factory** [reply]
Awesome, it [works now](https://pyright-playground.decorator-factory.su/?gzip=H4sIAOfJsGIC_1WPwWrDMBBE7_qKwb1Y4OQDAukxvYheGnIpRajOqhFIWrOSTfP3ldOGkrksyzxmGC-cUK9TyF8IaWKpOF4nOjkZ8EKZJIzK_zOWvivlEjiXO25CJXHxrUrzlTLY3xP6znQDPnnO5_0DpZVSY3Sl4MDc_9W8mw-9U2g6k4e1IYdqbV8o-gGLizPtYDQ2z3jlTL_kqhXY3vzWfLst3TO3b03vLil1Wgkt5KJtK6hvpgaekHghjLMUFlxISP0AtecW4w0BAAA%3D)!

It's still inferred as `Foo[str]` in mypy. Do you think it's worth making a discussion about this on `python/typing` or `python/mypy`?

**Reply by erictraut** [reply]
I would file a bug report with mypy.
